sovrin_node/persistence/idr_cache.py

```diff
@@ -0,0 +1,201 @@
+import os
+
+from collections import OrderedDict
+import rlp
+
+from plenum.common.constants import VERKEY, TRUSTEE, STEWARD
+from plenum.common.types import f
+from plenum.persistence.kv_store_leveldb import KVStoreLeveldb
+from sovrin_common.constants import ROLE, TGB, TRUST_ANCHOR
+from stp_core.common.log import getlogger
+
+logger = getlogger()
+
+
+class IdrCache:
+    """
+    A cache to store a role and verkey of an identifier, the db is only used to
+    store committed data, uncommitted data goes to memory
+    The key is the identifier and value is a pack of fields
+    The first byte indicates whether the identifier has a guardian or not,
+    if it has then the next few bytes will be the guardian's identifier
+    otherwise they will be the verkey. Then there is delimiter byte after which
+    the value of role starts. eg.
+    Value in case of guardian: '\2<guardian's DID>\0<role of the DID>'
+    Value in case of no guardian: '\1<verkey of the DID>\0<role of the DID>'
+    """
+
+    unsetVerkey = b'-'
+
+    def __init__(self, basedir: str, name):
+        logger.debug('Initializing identity cache {} at {}'
+                     .format(name, basedir))
+        self._basedir = basedir
+        self._name = name
+        self._db = None
+        self.open()
+        # OrderedDict where key is the state root after batch and value is a
+        # dictionary similar to cache which can be queried like the
+        # database, i.e `self._db`. Keys (state roots are purged) when they
+        # get committed or reverted.
+        self.unCommitted = OrderedDict() # type: Dict[bytes, OrderedDict]
+
+        # Relevant NYMs operation done in current batch, in order
+        self.currentBatchOps = []   # type: List[Tuple]
+
+    @staticmethod
+    def encodeVerkey(verkey):
+        if verkey is None:
+            return IdrCache.unsetVerkey
+        else:
+            if isinstance(verkey, str):
+                return verkey.encode()
+            return verkey
+
+    @staticmethod
+    def decodeVerkey(verkey):
+        if verkey == IdrCache.unsetVerkey:
+            return None
+        else:
+            return verkey.decode()
+
+    @staticmethod
+    def getPrefixAndIv(guardian=None, verkey=None):
+        iv = guardian if guardian else verkey
+        prefix = b'1' if guardian else b'0'
+        return prefix, iv
+
+    @staticmethod
+    def packIdrValue(ta=None, role=None, verkey=None):
+        # prefix, iv = IdrCache.getPrefixAndIv(guardian, verkey)
+        if ta is None:
```

Comment:
Please use the full name for variables. it's not a common abbreviation, and not clear for the reader of the code